File: 2023/day-6/day-6.py
with open('input.txt', 'r') as f:
    lines = f.readlines()

##################
###   Part 1   ###
##################

race_times = list(map(int, filter(None, lines[0].strip().split(" ")[1:])))
race_dists = list(map(int, filter(None, lines[1].strip().split(" ")[1:])))

races = zip(race_times, race_dists)

ans = 1

for race_time, race_dist in races:
    ways_to_beat = 0

    for time in range(race_time):
        speed = time
        moving_time = race_time - time

        if speed * moving_time > race_dist:
            ways_to_beat += 1

    ans *= ways_to_beat

print(f"Part 1: {ans}")

##################
###   Part 2   ###
##################

race_time = int(lines[0].strip().split(":")[1].replace(" ", ""))
race_dist = int(lines[1].strip().split(":")[1].replace(" ", ""))

start_beat = 0

for time in range(race_time):
    speed = time
    moving_time = race_time - time

    if speed * moving_time > race_dist:
        start_beat = time
        break

print(f"Part 2: {race_time - (start_beat * 2) + 1}")

File: 2023/day-6/day-6.rs
use std::fs::File;
use std::io::{ BufReader, BufRead };

fn main() {
    let file = File::open("input.txt").expect("Input file missing!");
    let reader = BufReader::new(file);
    let lines: Vec<String> = reader.lines().collect::<Result<_, _>>().unwrap();

    //////////////////
    //    Part 1    //
    //////////////////

    let race_times: &Vec<i32> = &lines[0].trim().split(":").collect::<Vec<_>>()[1].split(" ").collect::<Vec<_>>().iter().filter(|&time| time != &"").map(|time| time.parse::<i32>().unwrap()).collect();

    let race_dists: &Vec<i32> = &lines[1].trim().split(":").collect::<Vec<_>>()[1].split(" ").collect::<Vec<_>>().iter().filter(|&time| time != &"").map(|time| time.parse::<i32>().unwrap()).collect();

    let races = race_times.iter().zip(race_dists.iter());

    let mut answer = 1;

    for (race_time, race_dist) in races {
        let mut ways_to_beat = 0;

        for charge_time in 0..*race_time {
            let speed = charge_time;
            let moving_time = race_time - charge_time;

            if speed * moving_time > *race_dist {
                ways_to_beat += 1;
            }
        }

        answer *= ways_to_beat;
    }

    println!("Part 1: {}", answer);

    //////////////////
    //    Part 2    //
    //////////////////

    let race_time: i64 = lines[0].trim().split(":").collect::<Vec<_>>()[1].replace(" ", "").parse::<i64>().unwrap();
    let race_dist: i64 = lines[1].trim().split(":").collect::<Vec<_>>()[1].replace(" ", "").parse::<i64>().unwrap();

    let mut first_beat: i64 = 0;

    for charge_time in 0..race_time {
        let speed = charge_time;
        let moving_time = race_time - charge_time;

        if speed * moving_time > race_dist {
            first_beat = charge_time;
            break;
        }
    }

    println!("Part 2: {}", race_time - ( first_beat *  2 ) + 1);
}
